### PR TITLE
100 Login screen fixes

### PR DESCRIPTION
**Pull request by mindfulme**
#### What's this PR do?
Applies the other conditional rendering 
#### Where should the reviewer start?
At views/Login/Login.vue
#### How should this be manually tested?
Login via Private Key
#### Any background context you want to provide?
The pin code fields hidden by default, key file input disappears after entering brainkey, password field appears after uploading key file
#### What are the relevant tickets?
#100 
#### Screenshots (if appropriate)
screen shot 2018-10-22 at 15 21 46

#### URL
Provide a link to your deploy from http://staging-ui.trusty.apasia.tech:8080/
http://staging-ui.trusty.apasia.tech:8080/100-login-screen-fixes
#### Questions:
There is a place for @focus and @blur events, should UX be improved using them or this is enough (current conditions for the rendering seem ok too)?

### CODE REVIEW

**Comment by roma219** [comment]
Doesn't seem to fix everything that is stated in issue. Open when when fixes & ready for review

**Comment by roma219** [comment]
@MindfulMe is this WIP or done?

**Comment by mindfulme** [comment]
I am working on it right now

**Comment by mindfulme** [comment]
@roma219 I will ask for a review in 1 hour

**Comment by roma219** [comment]
@MindfulMe Sure, but next time please open PR when it is actually should be reviewed, or it becomes confusing

**Comment by mindfulme** [comment]
@roma219 got it

**Comment by mindfulme** [comment]
@roma219 though, by default input field deletes all spaces. So, I am wondering how backup phrase look like? Maybe, it is concatenated words into one big string without whitespaces?

**Comment by roma219** [comment]
@MindfulMe It does not delete all spaces, only trims trailing spaces at the end
image

File: src/views/Login/Login.vue [hunk 1]
```diff
@@ -131,6 +133,15 @@ export default {
             required: (value) => {
               if (this.file) return true
               return required(value)
+            },
+            brainkeyValidator: (value) => {
+              if (value.split(' ').length - 1 === 15) {
+                this.brainkeyCompleted = true
```
Comment (roma219) [inline src/views/Login/Login.vue]: The whole validator can simly be `return value.split(' ').length - 1 >= 15`, no need for additional variables and conditions. You can access validator's valid/invalid state throught `this.$v` object in component (see vuelidate docs for info).

File: src/helpers/inputMessages.js [hunk 2]
```diff
@@ -25,7 +26,8 @@ export default {
     title: 'backup phrase',
     tip: 'Enter 16 words backed up when account was created',
     validation: {
-      required: 'Enter backup phrase'
+      required: 'Enter backup phrase',
+      minLength: 'Must contain 16 words'
```
Comment (roma219) [inline src/helpers/inputMessages.js]: there is no minlength validator on a brainkey, isn't it?

**Review by roma219 — CHANGES_REQUESTED** [review]
Not reviewing the rest since it doesn't work right:
image
No error message about length of brainkey, login button not blocked, keyfile button appeared.

Hint: The whole issue might be too much for you know. I propose you start from scratch and fix just something from the issue, say only 1 point. State what you fixed in PR, i'll review it & merge, and than you can start working on another small thing. You need to understand how the whole validation mechanic works in Login/Signup

**Comment by mindfulme** [comment]
@roma219 I can cope with it

File: src/helpers/inputMessages.js (hunk 2 above)
Comment (mindfulme) [reply src/helpers/inputMessages.js]: yep, should be not `brainkeyValidator `,  rather `minLength`

File: src/views/Login/Login.vue (hunk 1 above)
Comment (mindfulme) [reply src/views/Login/Login.vue]: I think I can't call then minLength anywhere without passing a parameter `value` to it

File: src/helpers/inputMessages.js (hunk 2 above)
Comment (roma219) [reply src/helpers/inputMessages.js]: MinLength only checks for string length, not applicable to brainkey

File: src/views/Login/Login.vue [hunk 3]
```diff
@@ -71,7 +73,7 @@
 
       <div class="login__form">
         <Button
-          :disabled="loginDisabled"
+          :disabled="loginDisabled || (!file && !brainkeyCompleted && !password)"
```
Comment (roma219) [inline src/views/Login/Login.vue]: Should not be any other checks here, besides`loginDisabled`, move your condition to that computed value

File: src/views/Login/Login.vue (hunk 1 above)
Comment (roma219) [reply src/views/Login/Login.vue]: > I think I can't call then minLength anywhere without passing a parameter value to it

What it has to do with my comment?

**Review by roma219 — CHANGES_REQUESTED** [review]
Fix what is written in comments before submitting for review again

**Comment by roma219** [comment]
@MindfulMe Also remove the trimming of the last space by input in brainkey field

**Comment by mindfulme** [comment]
@roma219 refactored exactly according to your comments

File: src/views/Login/Login.vue [hunk 4]
```diff
@@ -131,7 +133,8 @@ export default {
             required: (value) => {
               if (this.file) return true
               return required(value)
-            }
+            },
+            validator: (value) => (value.split(' ').length - 1 >= 15)
```
Comment (roma219) [inline src/views/Login/Login.vue]: It is the same as naming function `function`. You had an ok name before.

File: src/views/Login/Login.vue [hunk 5]
```diff
@@ -207,6 +211,16 @@ export default {
     },
     removeFile() {
       this.file = null
+    },
+    onFocus() {
```
Comment (roma219) [inline src/views/Login/Login.vue]: `onFocus` of what? There are several fields in component. Rename it to onBrainkeyInputFocus

File: src/views/Login/Login.vue [hunk 6]
```diff
@@ -207,6 +211,16 @@ export default {
...
+      if (!this.$v.brainkey.validator && !this.file) {
+        this.showFileField = false
+      }
+    },
+    onBlur() {
```
Comment (roma219) [inline src/views/Login/Login.vue]: Same as onFocus

**Review by roma219 — CHANGES_REQUESTED** [review]
Better, needs a few changes. Will check thoughtfully later.

File: src/views/Login/Login.vue (hunk 6 above)
Comment (mindfulme) [reply src/views/Login/Login.vue]: just renamed to onBrainkeyInputBlur

File: src/views/Login/Login.vue (hunk 5 above)
Comment (mindfulme) [reply src/views/Login/Login.vue]: just renamed to onBrainkeyInputFocus

**Comment by roma219** [comment]
@MindfulMe This is not done:

-show KEY FILE button back when user exits the field (clicks cross or clicks outside)[after brainkey fully entered]
on Signup screen on tab Password change text "Your account will be secured only with password" to --"Faster, but less secure way to Signup. Your account will be secured only with password. Slower way to signup will better protect you"
-with yellow font in the row just under the line of the field Confirm password insert text "This password can't be recovered, write it on paper"
-on Signup screen on tab Private key change text "You need to backup account after creating a PIN code" to "Slower, but more secure way to Signup protects your account with a combination of PIN and Backup phrase. You need to backup account after creating a PIN code"

**Comment by mindfulme** [comment]
@roma219 The text has been changed several days ago, the other conditions were completed according to the issue ticket #100. The following will be re-written according to your comment above:

-show KEY FILE button back when the user exits the field (clicks cross or clicks outside)[after brainkey fully entered]

-with yellow font in the row just under the line of the field Confirm password insert text "This password can't be recovered, write it on paper"

**Comment by roma219** [comment]
@MindfulMe conflicts, merge latest dev this repo and vuex-bitshares

File: src/views/Login/Login.vue [hunk 7]
```diff
@@ -131,7 +130,8 @@ export default {
             required: (value) => {
               if (this.file) return true
               return required(value)
-            }
+            },
+            brainkeyValidator: (value) => (value.split(' ').length - 1 >= 15)
```
Comment (roma219) [inline src/views/Login/Login.vue]: just `value =>`

File: src/helpers/inputMessages.js [hunk 8]
```diff
@@ -18,14 +18,15 @@ export default {
   confirmPassword: {
     title: 'confirm password',
     validation: {
-      sameAsPassword: 'Passwords do not match'
+      sameAsPassword: 'Passwords do not match \n '
```
Comment (roma219) [inline src/helpers/inputMessages.js]: Why are you adding `\n`? There is no need for it

File: src/views/Login/Login.vue [hunk 9]
```diff
@@ -148,12 +148,13 @@ export default {
...
       type: 'password',
-      file: null
+      file: null,
+      showFileField: true
     }
   },
   computed: {
     loginDisabled() {
-      return this.$v.$dirty && this.$v.$invalid
+      return (this.$v.$dirty && this.$v.$invalid) || (!this.file && !this.$v.brainkey.brainkeyValidator && !this.password)
```
Comment (roma219) [inline src/views/Login/Login.vue]: What does this condition mean? This should be validated throught vuelidate `validations` config

**Review by roma219 — CHANGES_REQUESTED** [review]
Needs explanations and fixes

File: src/views/Login/Login.vue (hunk 7 above)
Comment (mindfulme) [reply src/views/Login/Login.vue]: could be just `value` - yes

File: src/views/Login/Login.vue (hunk 9 above)
Comment (mindfulme) [reply src/views/Login/Login.vue]: I found 3 ways to complete it, okay - I will use vuelidate validations config

File: src/views/Login/Login.vue (hunk 7 above)
Comment (mindfulme) [reply src/views/Login/Login.vue]: thought in `required` could be just `value`

**Comment by mindfulme** [comment]
@roma219 review & merge if ok

**Comment by roma219** [comment]
@MindfulMe not ok

File: src/views/Login/Login.vue [hunk 10]
```diff
@@ -119,7 +120,9 @@ export default {
         return {
           name: {},
           password: {},
-          brainkey: {},
+          brainkey: {
+            isLoginDisabled: () => ((this.$v.$dirty && this.$v.$invalid) || (!this.file && !this.$v.brainkey.brainkeyValidator && !this.password))
```
Comment (roma219) [inline src/views/Login/Login.vue]: wtf is this

File: src/views/Login/Login.vue [hunk 11]
```diff
@@ -128,10 +131,12 @@ export default {
...
           password: {},
           brainkey: {
-            required: (value) => {
+            required: value => {
               if (this.file) return true
               return required(value)
-            }
+            },
+            brainkeyValidator: value => (value.split(' ').length - 1 >= 15),
+            isLoginDisabled: () => ((this.$v.$dirty && this.$v.$invalid) || (!this.file && !this.$v.brainkey.brainkeyValidator && !this.password))
```
Comment (roma219) [inline src/views/Login/Login.vue]: what is this

**Comment by roma219** [comment]
@MindfulMe read this fully first, you need to understand what you are doing
https://monterail.github.io/vuelidate/

**Comment by mindfulme** [comment]
@roma219 got it

**Comment by mindfulme** [comment]
@roma219 I completely misunderstood you, I did a rollback and removed the `this.$v.$dirty && this.$v.$invalid`

**Comment by mindfulme** [comment]
@roma219 review & merge if ok

File: src/views/Login/Login.vue (hunk 10 above)
Comment (mindfulme) [reply src/views/Login/Login.vue]: I had the same reaction when I saw you want me to change the validation, the thing was to remove the useless `this.$v.$dirty && this.$v.$invalid` which were here by default

Comment (mindfulme) [reply src/views/Login/Login.vue]: fixed by the latest commit

File: src/views/Login/Login.vue [hunk 12]
```diff
@@ -148,12 +148,13 @@ export default {
...
       inProgress: false,
       type: 'password',
-      file: null
+      file: null,
+      showFileField: true
     }
   },
   computed: {
     loginDisabled() {
-      return this.$v.$dirty && this.$v.$invalid
```
Comment (roma219) [inline src/views/Login/Login.vue]: Why are you removing this? This simply disabled button when vuelidate validation doesn't pass, so this should remain how it was. You need to add your checks to validation configs itself.

**Review by roma219 — CHANGES_REQUESTED** [review]

**Comment by roma219** [comment]
> @roma219 I completely misunderstood you, I did a rollback and removed the `this.$v.$dirty && this.$v.$invalid`

You shouldn't remove it. You need to read through the docs first to understand how it works.

**Comment by mindfulme** [comment]
@roma219 I have thoroughly read Vuelidate docs also several posts on Monterail blog.  I have changed the default condition as it was not fully correct, and tested everything - now the button is enabled when required fields are filled & satisfy the requirements

File: src/views/Login/Login.vue [hunk 13]
```diff
@@ -148,12 +148,13 @@ export default {
...
+      return this.$v.$dirty || this.$v.$invalid
```
Comment (roma219) [inline src/views/Login/Login.vue]: It was && for specific reason - to initially show "clean" view without disabled button and only disable it after first validations ( = dirty checking)

**Review by roma219 — CHANGES_REQUESTED** [review]
Revert changes with $dirty checking (see comment) and it's good to go

File: src/views/Login/Login.vue (hunk 13 above)
Comment (mindfulme) [reply src/views/Login/Login.vue]: @roma219 done

**Review by roma219 — APPROVED** [review]

**Pull request merged**